warp/server.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/sirupsen/logrus"
+)
+
+// StartServer starts a warp proxy server with the given configuration.
+// It blocks indefinitely.
+func StartServer(cfg ServerConfig) error {
+	hostname, err := validation.ValidateHostname(cfg.Hostname)
+	if err != nil {
+		return err
+	}
+	if cfg.ClientID == "" {
+		cfg.ClientID = generateRandomClientID()
+	}
+
+	cfg.Tags = append(cfg.Tags, tunnelpogs.Tag{Name: "ID", Value: cfg.ClientID})
+
+	cfg.ServerURL, err = validation.ValidateUrl(cfg.ServerURL)
+	if err != nil {
+		return fmt.Errorf("validating server URL: %v", err)
+	}
+
+	// Check that the user has acquired a certificate using the log in command
+	originCertPath, err := homedir.Expand(cfg.OriginCert)
+	if err != nil {
+		return fmt.Errorf("cannot resolve path %s: %v", cfg.OriginCert, err)
+	}
+	ok, err := fileExists(originCertPath)
+	if !ok {
+		return fmt.Errorf(`Cannot find a valid certificate for your origin at the path:
+
+    %s
+
+If the path above is wrong, specify the path with the -origincert option.
+If you don't have a certificate signed by Cloudflare, run the command:
+
+    %s login
+`, originCertPath, os.Args[0]) // TODO - we need to improve how this is handled
+	}
+	// Easier to send the certificate as []byte via RPC than decoding it at this point
+	originCert, err := ioutil.ReadFile(originCertPath)
+	if err != nil {
+		return fmt.Errorf("cannot read %s to load origin certificate: %v", originCertPath, err)
+	}
+
+	tunnelMetrics := origin.NewTunnelMetrics()
+
+	httpTransport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   cfg.Timeout,
+			KeepAlive: cfg.KeepAlive,
+			DualStack: cfg.DualStack,
+		}).DialContext,
+		MaxIdleConns:          cfg.MaxIdleConns,
+		IdleConnTimeout:       cfg.IdleConnTimeout,
+		TLSHandshakeTimeout:   cfg.TLSHandshakeTimeout,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig:       &tls.Config{RootCAs: tlsconfig.LoadOriginCertsPool()},
+	}
+
+	tunnelConfig := &origin.TunnelConfig{
+		EdgeAddrs:         cfg.EdgeAddrs,
+		OriginUrl:         cfg.ServerURL,
+		Hostname:          hostname,
+		OriginCert:        originCert,
+		TlsConfig:         tlsconfig.CreateTunnelConfig(cfg.TLSConfig, cfg.EdgeAddrs),
+		ClientTlsConfig:   httpTransport.TLSClientConfig,
+		Retries:           cfg.Retries,
+		HeartbeatInterval: cfg.HeartbeatInterval,
+		MaxHeartbeats:     cfg.MaxHeartbeats,
+		ClientID:          cfg.ClientID,
+		ReportedVersion:   cfg.ReportedVersion,
+		LBPool:            cfg.LBPool,
+		Tags:              cfg.Tags,
+		HAConnections:     cfg.HAConnections,
+		HTTPTransport:     httpTransport,
+		Metrics:           tunnelMetrics,
+		MetricsUpdateFreq: cfg.MetricsUpdateFreq,
+		ProtocolLogger:    cfg.ProtoLogger,
+		Logger:            cfg.Logger,
+		IsAutoupdated:     cfg.IsAutoupdated,
+	}
+
+	// blocking
+	return origin.StartTunnelDaemon(tunnelConfig, cfg.ShutdownChan, cfg.ConnectedChan)
+}
+
+func fileExists(path string) (bool, error) {
```

Comment:
Is this function redundant with `if _, err := os.Stat(path); err == nil { ... }`?

Reply:
For the purposes of whether a file _exists_, yes, more or less -- although you'd want to check if `!os.IsNotExist(err)` instead of a blanket `err == nil`, since there might be other reasons that statting a file fails.